waspc/e2e-test/ShellCommands.hs

```diff
@@ -0,0 +1,83 @@
+
+-- TODO: Should we consider separating shell parts, Wasp CLI parts, and test helper parts in the future?
+--       This would likely reqiure adoption of some library, and creating new layers of abstraction.
+--       Deemed not worth it right now by all.
+
+-- TODO: In future, find a good way to test `wasp-cli start`.
+
+-- NOTE: Using `wasp-cli` herein so we can assume using latest `stack install` in CI and locally.
+
+type ShellCommand = String
+
+-- A way to create a shell command given some context info.
+type MakeShellCommand = ShellCommandContext -> ShellCommand
+
+-- Each shell command gets access to the current project name and current/golden output directories.
+data ShellCommandContext = ShellCommandContext
+  { currentProjectName :: String,
+    currentOutputDirAbsFp :: FilePath,
+    goldenOutputDirAbsFp :: FilePath
+  }
+  deriving (Show)
+
+combineShellCommands :: [ShellCommand] -> ShellCommand
+combineShellCommands = intercalate " && "
+
+combineMakeShellCommands :: [MakeShellCommand] -> MakeShellCommand
+combineMakeShellCommands commands context = combineShellCommands (map (\command -> command context) commands)
+
+cdIntoCurrentProject :: MakeShellCommand
+cdIntoCurrentProject context = "cd " ++ currentProjectName context
+
+appendToWaspFile :: String -> MakeShellCommand
+appendToWaspFile content _context =
+  -- NOTE: Using `show` to preserve newlines in string.
+  "echo " ++ show content ++ " >> main.wasp"
+
+waspCliNew :: MakeShellCommand
+waspCliNew context = "wasp-cli new " ++ currentProjectName context
+
+waspCliCompile :: MakeShellCommand
+waspCliCompile _context = "wasp-cli compile"
+
+-- TODO: We need to be careful what migration names we accept here, as Prisma will
+--       normalize a migration name containing spaces/dashes to underscores (and maybe other rules).
+--       This will impact our ability to move directories around if the names do not match exactly.
+--       Put in some check eventually.
+waspCliMigrate :: String -> MakeShellCommand
+waspCliMigrate migrationName _context =
+  let generatedMigrationsDir = joinPath [".wasp", "out", "db", "migrations"]
+      waspMigrationsDir = "migrations"
+   in combineShellCommands
+        [ -- Migrate using a migration name to avoid Prisma asking via CLI.
+          "wasp-cli db migrate-dev " ++ migrationName,
+          -- Rename both migrations to remove the date-specific portion of the directory to something static.
+          "mv " ++ (waspMigrationsDir </> ("*" ++ migrationName)) ++ " " ++ (waspMigrationsDir </> ("no-date-" ++ migrationName)),
+          "mv " ++ (generatedMigrationsDir </> ("*" ++ migrationName)) ++ " " ++ (generatedMigrationsDir </> ("no-date-" ++ migrationName))
+        ]
+
+waspCliBuild :: MakeShellCommand
+waspCliBuild _context =
+  combineShellCommands
+    [ -- Change DB to postgres by adding string at specific line so it still parses.
+      "awk 'NR==2{print \"  db: { system: PostgreSQL },\"}1' main.wasp > main.wasp.tmp",
```

Comment:
A bit scary that it does this, I wouldn't expect this based on the name.

I would consider renaming it to `setDbToPSQLAndRunWaspCliBuild`.

Or make it two separate commands.

Or maybe best, do this part directly in the definition of a golden test.

This part with setting db to PostgreSQL also looks very fragile, but ok, I understand that this was the fastest way to do it now. Maybe comment on it that it is very fragile, and assuming a lot of stuff.

Reply:
It is a bit bad that it is so hard for us to modify Wasp code.
But real solution is probably to not use `wasp new` and then patch it up but instead provide fresh wasp file just for this test. But ok, we can do this in the future.

Reply:
It is a bit bad that it is so hard for us to modify Wasp code.
But real solution is probably to not use `wasp new` and then patch it up but instead provide fresh wasp file just for this test. But ok, we can do this in the future.

Reply:
Good point, yes let me separate it out into two functions (both they have to explicitly call from a golden test) and note the fragility here. In general, I tried to keep the patches in code vs using filesystem stuff (a la fixtures/overlays). The fragility in the fresh Wasp file is if any syntax in there changes, or we add/remove stuff to default main.wasp, it too could "break" (but possibly silently, which may be worse?). I think for now with 2 separate commands we can prob live with the assumption if we put that postgres snippet in line 2 under the `app` decl, we are prob ok for now. But agree, not ideal under any scenario. :/

Reply:
Yeah, that sounds good enough for now!